### ISSUE TITLE
Determine if RBAC updates should drain

### ISSUE THREAD

**Issue by howardjohn** [istio/ztunnel#311]
Consider a long lived connection. After it is established, we create a rule that would have denied it. Should the connection be terminated, or should only new connections start being terminated?

Prior art:
* Istio: Drains
* Cilium: operates on many layers, so needs more research
* Ztunnel currently: does not drain

**Comment by keithmattix** [comment]
Definitely feels like it should drain IMO

**Comment by GregHanson** [comment]
this also impact waypoints.  If a long lived connection exists, and a waypoint is deployed, ztunnel will not direct traffic through the waypoint until the original connection is closed and re-established.

steps to reproduce using [online boutique](https://github.com/GoogleCloudPlatform/microservices-demo/tree/main) microservice:
```
istioctl install --set tag=1.18.0 -y --set profile=ambient --set meshConfig.accessLogFile=/dev/stdout
kubectl label namespace default istio.io/dataplane-mode=ambient --overwrite
kubectl apply -f https://raw.githubusercontent.com/GoogleCloudPlatform/microservices-demo/main/release/kubernetes-manifests.yaml

# deploy namespace wide waypoint
istioctl x waypoint apply -n default

# no access logs
kubectl logs deploy/namespace-istio-waypoint -f

kubectl delete pod -lapp=loadgenerator

# access logs generated per request as expected
kubectl logs deploy/namespace-istio-waypoint -f
```

**Comment by kyessenov** [comment]
I don't understand what "drain" here means. There's no general drain strategy for TCP connections. You can only abruptly terminate them. Is that what's proposed?

**Comment by howardjohn** [comment]
Yes, drain was poorly worded. Although fwiw I'm not proposing it, just that
we make an explicit decision - I am leaning towards not terminating

**Comment by kyessenov** [comment]
I think never terminating that would be dangerous in an eventually consistent system. When workloads and policies are deployed concurrently (enabling waypoint is a policy), the order should not matter when things open connections. So it seems necessary to terminate eventually, and the challenge is defining when to terminate connections and limiting the scope to only those should be.

**Comment by linsun** [comment]
Agreed, I'd expect updated RBAC become effective even with long lived connection.

cc @louiscryan for input

**Comment by howardjohn** [comment]
I agree with the concerns of eventual consistency. For RBAC it makes more sense I think - we explicitly would close the connection if it was new. For Waypoint being added, though, closing the connection seems unneccesarily aggressive

**Comment by bleggett** [comment]
> I agree with the concerns of eventual consistency. For RBAC it makes more sense I think - we explicitly would close the connection if it was new. For Waypoint being added, though, closing the connection seems unneccesarily aggressive.

Dropping a connection that would previously have been allowed but no longer is, as soon as we are aware of that, is an non-negotiable obligation on our end, IMO.

Scoping that to "pure RBAC" so we don't have to do the same when L7 concerns are added with Waypoints is both a complex behavior to actually model, and creates behavioral exceptions in ztunnel around connection handling that make it harder to reason about.

I think we just drop in both cases, because that's the simplest, safest, and most obvious behavior.

**Comment by kyessenov** [comment]
I think we're over-indexing on TCP handling in ztunnel. If it was working on the packets, it would be completely natural to drop packets as soon as the routing + rbac imply so. Since ztunnel has to support UDP and L3 eventually, it is the necessary thing to do. The fact that errors are opaque is a general problem with non-application level protocols.

**Comment by howardjohn** [comment]
AFAIK most CNIs actually do not drop packets when the netpol changes. At
least that has been my experience a few years ago - worth surveying the
landscape

**Comment by hzxuzhonghu** [comment]
I am not sure how can we be ware of the rbac failure after a HBONE request was sent, IIUC the authz is only done once for a tcp connection

**Comment by ilrudie** [comment]
It seems there are (potentially) 2 difference scenarios here:

1. Change to traffic flow is requested, i.e. they added a waypoint. In this scenario traffic which is bypassing the waypoint may be something L4 is obliged to stop.
2. Change to existing L4 policy. In this case we should probably try to evaluate whether or not existing connections violate the new policy. If they do, we're likely obliged to halt it, if they are still allowed we don't need to do anything.

I think we're also able to say that item 1 isn't actually a real thing. Existence of a waypoint doesn't imply in anyway that traffic strictly must use it. User's should express this via auth API and thus item 2 is really the only problem we're concerned with. It could potentially be a property exposed via the SuperService whether or not we manage these policies for the user.

**Comment by howardjohn** [comment]
> Change to existing L4 policy. In this case we should probably try to evaluate whether or not existing connections violate the new policy. If they do, we're likely obliged to halt it, if they are still allowed we don't need to do anything.

I definitely agree that _if_ we change something, we should not do anything if the new policies are allowed (only if they deny, then we close). So we would not want a traditional Envoy "drain".

We still need to decide if we want to apply them to existing connections. I think we should

**Comment by linsun** [comment]
@ilrudie ok to assign this to you?  Can we identify action item for this in terms of beta for L4 first?  Seems everyone agrees that we should do something when RBAC updates, even for long lived connections.

**Comment by ilrudie** [comment]
@linsun, Sure! Happy to work on this.

**Comment by ilrudie** [comment]
Added a topic to the ambient WG to try to finalize consensus here before diving into implementation.

**Comment by hzxuzhonghu** [comment]
My opinion is for L7 we can deny future request, but for L4 there is no need, and also hard because the authz is applied per connection, while it is per request for L7

**Comment by bleggett** [comment]
> Existence of a waypoint doesn't imply in anyway that traffic strictly must use it.

It does imply that ztunnel must use the waypoint, however. 

I'm cautious about involving ztunnel in any L7 policy decision more complex than "this connection must go thru a waypoint, YES/NO".

**Comment by ilrudie** [comment]
@hzxuzhonghu To be clear: you're saying you do want to close/drain existing connections, yeah?

**Comment by hzxuzhonghu** [comment]
Not exactly, as to L4, the rbac is enforced once after connection established. We donot have the chance to close connection.

For L7, as the connection is shared, the authz is for requests, there is no reason to close the connection.

**Comment by ilrudie** [comment]
High level sketch, what I'm looking at right now is having a proxy instance be able to handle it's own connections. We'll either poll for auth updates or have a channel where something like a policy update can include notifying the routine that will close these channels (undecided). Each connection will have a channel it can check which will cause it to close and those close channels will likely be keyed by rbac::Connection in a shared hashmap so the proxy has a relatively easy time of determining which connections must be closed. It's pretty WIP but hope to have some code to peak at soon.

**Comment by hzxuzhonghu** [comment]
But each connection we do assert_rbac once now.

**Comment by ilrudie** [comment]
You're correct. At present we only assert_rbac once. The proposal is to have a way for the zt to assert_rbac on existing connections and then close/drain them via a channel. It's net new code/functionality for zt.

**Comment by ilrudie** [comment]
Sorry it'll be a bit tougher to read since it's built on top of zt code which hasn't merged yet but #772 has a partial implementation of what I was thinking about in terms of being able to assert policy against connections after they were established.

**Comment by mikemorris** [comment]
I've seen the behavior of _not_ re-authorizing long-lived connections when AuthZ policy changes confuse users wondering why their rules don't seem to be effective - this is something we had [wanted to implement in Consul](https://github.com/hashicorp/consul/issues/6454) but seemed to be prohibitively difficult without Envoy changes. The implementation @ilrudie is describing sounds substantially aligned to what I think the optimal user experience should be, excited to see progress on this!

**Comment by kyessenov** [comment]
I think you need to support draining at the tunnel protocol level to make enforcement tolerable in terms of UX. What I mean is that a connection that is destined to be terminated should signal the peers that are application aware to terminate it gracefully. Ztunnel cannot do it for HTTP but if it uses a waypoint, an HTTP connection to ztunnel should not be abruptly closed by ztunnel without asking waypoint to send a GO_AWAY frame. Otherwise, this is a classical 503 scenario and not acceptable in production.

**Comment by ilrudie** [comment]
It seems to me that layer targeted policy would be helpful here. If you target L4 then it's explicitly not application enforced and DROP seems like suitable behavior. If you want some sort of gentle application-aware handling then target L7 layer with your policy.

**Comment by bleggett** [comment]
> It seems to me that layer targeted policy would be helpful here. If you target L4 then it's explicitly not application enforced and DROP seems like suitable behavior. If you want some sort of gentle application-aware handling then target L7 layer with your policy.

+ 1 and we should already get that in most cases IIUC - if a policy is applied that should terminate a L7 connection, waypoint should terminate it, and ZT is none the wiser. 

If you nuke connections at the L4 layer, regardless of whether the connection might be carrying a higher-level protocol, there's little we can do but drop it.

I guess we could set up a backchannel to the waypoint to inform it and give it time to do something, but that adds complexity and I don't see it as a beta blocker.

**Comment by ilrudie** [comment]
> I guess we could set up a backchannel to the waypoint to inform it and give it time to do something, but that adds complexity and I don't see it as a beta blocker.

Yeah, it seems plausible to make but it would likely wind up a special behavior of the reference implementation that requires Istio ztunnel and waypoint. I'd agree re: Beta and add that having a good methodology for how to try to achieve this type of behavior without needing tightly coupled components is probably a strong start.

**Comment by kyessenov** [comment]
Yeah, I think it's pretty clear that you _can't_ drain L4 RBAC. So the answer to the issue title is "no", and the choice is how often to enforce policies to drop (at start or periodically).

**Comment by ilrudie** [comment]
If anyone was holding off on taking a look because the diff was huge I've rebased against master now that Yuval's inpod has merged and it's much more reasonable manageable.

**Comment by zengyuxing007** [comment]
Istio's L4 AuthorizationPolicy functionality essentially overlaps with  the Kubernetes [networkpolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/#networkpolicy-s-impact-on-existing-connections). 
NetworkPolicy usually does not process existing connections, but only takes effect for new connections.

**Comment by zengyuxing007** [comment]
https://github.com/istio/istio/issues/48876  Similar issue at the beginning of Ambient deploy/open.

**Comment by ilrudie** [comment]
It seems fundamentally different to me. In your pr the intent is to allow not severing existing connections when adding a wl to ambient. One could say the intent is clear here and there are plenty of controls to manage this BUT adding to ambient isn't an explicitly "destructive" operation in the way denying connectivity is. In the case of auth policy change it would be allowing a connection the user explicitly denied to continue which seems odd since they have taken an action specifically to deny this connectivity. User intent seems way less ambiguous when they've said "deny this connection".

Louis's comments about disconnecting cause from effect also hold a lot of sway here IMO. If the user made a mistake it's better they find out when they make it vs finding out at some indeterminate point in the future. We do them no favors by allowing them to accidentally make a ticking time bomb.

**Comment by linsun** [comment]
Hi @ilrudie - should we close this? I recall you implemented this already?

**Comment by ilrudie** [comment]
@linsun - yeah, closed by #772
